src/make_watertight/Gen.cpp

```diff
@@ -0,0 +1,2117 @@
+  }
+  return retval;
+}
+
+moab::ErrorCode Gen::delete_surface(moab::EntityHandle surf, moab::Tag geom_tag, moab::Range tris, int id, bool debug, bool verbose) {
+
+  moab::ErrorCode result;
+
+  //measure area of the surface
+  double area;
+  result = measure(surf, geom_tag, area, debug, verbose);
+  MB_CHK_SET_ERR(result, "could not measure area");
+  assert(moab::MB_SUCCESS == result);
+
+  //remove triagngles from the surface
+  result = MBI()->remove_entities(surf, tris);
+  MB_CHK_SET_ERR(result, "could not remove tris");
+  assert(moab::MB_SUCCESS == result);
+  //print information about the deleted surface if requested by the user
+  if (debug)
+    std::cout << "  deleted surface " << id
+              << ", area=" << area << " cm^2, n_facets=" << tris.size() << std::endl;
+
+  // delete triangles from mesh data
+  result = MBI()->delete_entities(tris);
+  MB_CHK_SET_ERR(result, "could not delete tris");
+  assert(moab::MB_SUCCESS == result);
+
+  //remove the sense data for the surface from its child curves
+  result = remove_surf_sense_data(surf, debug);
+  MB_CHK_SET_ERR(result, "could not remove surface's sense data");
+
+  //if this was the last surface in a volume, then delete the volume
+  std::vector<moab::EntityHandle> parent_volumes;
+  result = MBI()->get_parent_meshsets(surf, parent_volumes);
+  MB_CHK_SET_ERR(result, "could not get the surface's parent meshsets");
+  assert(moab::MB_SUCCESS == result);
+
+  //check each parent volume
+  std::vector<moab::EntityHandle>::iterator i;
+  for (i = parent_volumes.begin(); i != parent_volumes.end(); i++) {
+    moab::EntityHandle parent_vol = *i;
+    std::vector<moab::EntityHandle> child_surfs;
+    result = MBI()->get_child_meshsets(parent_vol, child_surfs);
+    MB_CHK_SET_ERR(result, "could not get the child surfaces of the volume");
+    assert(moab::MB_SUCCESS == result);
+
+    // delete volume if it only has this surface as its child
+    if (child_surfs.size() == 1 && child_surfs[0] == surf) {
+      result = delete_vol(parent_vol);
+      MB_CHK_SET_ERR(result, "could not delete the parent volume");
+    }
+  }
+
+  //remove the surface set itself
+  result = MBI()->delete_entities(&(surf), 1);
+  MB_CHK_SET_ERR(result, "could not delete surface set");
+  assert(moab::MB_SUCCESS == result);
+
+
+  return moab::MB_SUCCESS;
+}
+
+/// removes sense data from all curves associated with the surface given to the function
+moab::ErrorCode Gen::remove_surf_sense_data(moab::EntityHandle del_surf, bool debug) {
+
+  moab::ErrorCode result;
+  moab::GeomTopoTool gt(MBI(), false);
+  int edim = gt.dimension(del_surf);
+
+  if (edim != 2) {
+    MB_CHK_SET_ERR(moab::MB_FAILURE, "could not remove sense data: entity is of the wrong dimension");
+  }
+
+// get the curves of the surface
+  moab::Range del_surf_curves;
+  result = MBI() -> get_child_meshsets(del_surf, del_surf_curves);
+  MB_CHK_SET_ERR(result, "could not get the curves of the surface to delete");
+  if (debug)
```

Comment:
Simplicity: This `if` condition can be collapsed with the one starting on line 1764.